### ISSUE TITLE
Unit testing error resolving class

### ISSUE THREAD

**Issue by stratus1011** [apollographql/apollo-kotlin#4669]
### Version

3.7.4

### Summary

In relation to this issue from December: https://github.com/apollographql/apollo-kotlin/issues/4568

I am having the same problem in a multi-module project where if you change anything in a Unit test and then re-run that test it will fail with the error until you do a clean build and then run it again.

`Cannot access class '<classname>'. Check your module classpath for missing or conflicting dependencies`

I have narrowed it down to whether or not we specify a "service" in our gradle file or not.

If we **don't** set a service like this everything works:
```kotlin
apollo {
     packageName.set("com.example.package")
     addTypename.set("always")
     generateDataBuilders.set(true)
}
```

If we **do** set a service it consistently has the behavior described above:
```kotlin
apollo {
     service("myservice"){
          packageName.set("com.example.package")
          addTypename.set("always")
          generateDataBuilders.set(true)
     }
}
```

We recently made the change to specify the service since that is the recommended approach going forward, which is how I was able to correlate the change with the issue.

### Steps to reproduce the behavior
Create a multi-module project and set a "service" configuration.

```kotlin
apollo {
     service("myservice"){
          packageName.set("com.example.package")
          addTypename.set("always")
          generateDataBuilders.set(true)
     }
}
```

### Logs

```
(Your logs here)
```

**Comment by BoD** [comment]
Hi!
At the moment, I couldn't reproduce with a very basic project that has several modules and services, [here](https://github.com/BoD/MultiModuleAndServiceApolloAndroid3). Are you able to reproduce it with that project? If not, it would help if you could try to find and point to differences between your own project and this one, so we have a chance to reproduce and understand the issue. Thanks!

**Comment by stratus1011** [comment]
I'll take a look at the project you referenced.

**Comment by stratus1011** [comment]
@BoD I wasn't able to produce it on the project you gave me in the exact format it was in. The project I am working on is an Android project and it fails on this gradle task `compileDebugUnitTestKotlin`. I took the project you sent me and converted it into an Android project and immediately began getting the issue. The specific use case I was testing was creating a graphql Fragment and creating an extension method on it. I got this error.

```
e: /HelloTest.kt: (25, 39): Cannot access class 'com.example.servicea.fragment.ScreenFragment'. Check your module classpath for missing or conflicting dependencies
```

So, the only difference in setup I was able to find was the fact that it was an Android project. I mimicked the exact structure the sample project had, but the modules were Android Library modules instead. I hope that helps narrow down what might be happening. It could have to do with the Android variants and resolving the correct class when compiling.

**Comment by BoD** [comment]
I made [this repository](https://github.com/BoD/apollo-kotlin-template-multi-modules-and-services-android) and I had one instance where `SchemaModuleServiceAQuery` was appearing red as not resolved in AS, but building on the command line was working. Then I noticed that `graphqlSchema/build/generated/source/apollo/service-a` didn't have the "sources root" label next to it (see screenshot where the label is present on both `service-a` and `service-b`).

<img width="327" alt="Screenshot 2023-02-07 at 11 11 24" src="https://user-images.githubusercontent.com/372852/217216184-a15bf91f-060f-47d4-a6ec-57a19b9ae0d7.png">

I clicked on "reload gradle project" and that fixed the issue, and I'm no longer able to reproduce it at the moment, even when restarting AS / re-importing the project / running unit tests...

On your side, are you able to reproduce the issue in a systematic way with this project?

**Comment by stratus1011** [comment]
The issue comes when I was using a graphql Fragment and then creating an extension method on that Fragment. I have an example repo that is similar to yours. You don't need to run the code, just run the unit tests in the `HelloTest.kt` file. https://github.com/stratus1011/apollo-example. Run the test once and it will probably work. Change something in the test and try to run it again and you'll see the classpath issue I've referred to. A clean build makes the test work again.

<img width="296" alt="image" src="https://user-images.githubusercontent.com/6954348/217281392-8a808636-0d5e-47f2-a8ba-e9f0c97b2096.png">

**Comment by BoD** [comment]
Thanks a lot, I can indeed reproduce it! Minimal repro project [here](https://github.com/BoD/apollo-kotlin-template-multi-modules-and-services-android/tree/issue-4669).
For now the precise source of the issue is unclear, but seems to be related to the method used to register the codegen task with the Android plugin, [which is different](https://github.com/apollographql/apollo-kotlin/blob/f6cbc453e8728a9ad576e6635fa329b18849cd39/libraries/apollo-gradle-plugin-external/src/main/kotlin/com/apollographql/apollo3/gradle/internal/DefaultApolloExtension.kt#L558) for the default service.

Also not exactly sure why it's a problem with extensions methods only (as far as I can see).

**Comment by martinbonnin** [comment]
I filed a bug in Android Gradle Plugin [there](https://issuetracker.google.com/issues/268218176)

@stratus1011 in the meantime, you can specify the `outputDirConnection` explicitely as a workaround:

```kotlin
apollo {
    service("service") {
        packageName.set("com.example")
        outputDirConnection {
            connectToAndroidSourceSet("main")
        }
    }
}
```

**Comment by BoD** [comment]
A fix has been released as part of [v3.7.5](https://github.com/apollographql/apollo-kotlin/releases/tag/v3.7.5)

**Comment by nateridderman** [comment]
This bug affected my project while using 3.7.4, but it was still broken after upgrading to 3.7.5 or even 3.8.1. The workaround worked however.  

I'm using AGP 8.2.0-rc03 and Kotlin 1.7.0
